[don't merge] handle a potential zero denominator in a substitution in _meijerint_indefinite_1

ehren opened the pull request:
towards a solution of integrate((x**(n-1))*sqrt(1+x**n), x) not returning a Piecewise with the n = 0 case.

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

doesn't quite address #5949

#### Brief description of what is fixed or changed


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->

<!-- END RELEASE NOTES -->

ehren committed:
handle a potential zero denominator in a substitution in _meijerint_indefinite_1

towards a solution of integrate((x**(n-1))*sqrt(1+x**n), x) not
returning a Piecewise with the n = 0 case.

ehren commented:
Still tinkering. This branch now has a similar issue as the existing behaviour of `integrate(x**n, x, meijerg=True)`

```
>>> integrate((x**(n-1))*sqrt(1+x**n), x)
Piecewise((sqrt(2)*log(x), Eq(n, 0) & (Abs(x) < 1)), (-sqrt(2)*log(1/x), Eq(n, 0) & (1/Abs(x) < 1)), (-sqrt(2)*meijerg(((), (1, 1)), ((0, 0), ()), x) + sqrt(2)*meijerg(((1, 1), ()), ((), (0, 0)), x), Eq(n, 0)), (2*x**n*sqrt(x**n + 1)/(3*n) + 2*sqrt(x**n + 1)/(3*n), True))
```

asmeurer commented on sympy/integrals/meijerint.py:
```diff
@@ -1660,6 +1661,18 @@ def _meijerint_indefinite_1(f, x):
+        special_results = []
+        for den in (a, b):
+            if not den.is_Number and not den.is_nonzero:
+                zero_sol = solve(den, dict=True)
```
I'm not sure if we should worry about doing solve here. This seems like something that piecewise simplification should worry about. Do the other integration methods do this when they split into piecewise?

ehren replied:
I tried something like this instead of solve (diff against the first two commits):

```
diff --git a/sympy/integrals/meijerint.py b/sympy/integrals/meijerint.py
index 36609255ee..c72a78963f 100644
--- a/sympy/integrals/meijerint.py
+++ b/sympy/integrals/meijerint.py
@@ -1665,13 +1665,16 @@ def _meijerint_indefinite_1(f, x):
         special_results = []
         for den in (a, b):
             if not den.is_Number and den.is_zero is not False:
-                zero_sol = solve(den, dict=True)
-                for sol in zero_sol:
-                    zero_res = integrate(f.subs(sol), x)
-                    zero_cond = cond
-                    for vz, sz in sol.items():
-                        zero_cond = And(zero_cond, Eq(vz, sz))
-                    special_results.append(Piecewise((zero_res, zero_cond)))
+                _, den_mul_args = den.as_coeff_mul()
+                den = Mul(*den_mul_args)
+                f_replaced = f.xreplace({den: 0})
+                if f_replaced == f:
+                    # leave unevaluated
+                    zero_res = Integral(f_replaced, x)
+                else:
+                    zero_res = integrate(f_replaced, x)
+                zero_cond = And(cond, Eq(den, 0))
+                special_results.append(Piecewise((zero_res, zero_cond)))
 
         # we now use t**rho*G(params, t) = G(params + rho, t)
         # [L, page 150, equation (4)]

```

which works for this case:

```
>>> x, n, k = symbols("x, n, k") 
>>> integrate(sin(x**(n*k)), x)
Piecewise((x*sin(1), Eq(k*n, 0)), (x*x**(k*n)*gamma(1/2 + 1/(2*k*n))*hyper((1/2 + 1/(2*k*n),), (3/2, 3/2 + 1/(2*k*n)), -x**(2*k*n)/4)/(2*k*n*gamma(3/2 + 1/(2*k*n))), True))
```

but leaves the special case unevaluated with this example:

```
>>> integrate(sin(x**(n*k + 1)), x)
Piecewise((Integral(sin(x*x**(k*n)), x), Eq(2*k*n + 2, 0)), (x*x**(k*n)*x**(k*n/(k*n + 1))*x**(1/(k*n + 1))*gamma(1/2 + 1/(2*(k*n + 1)))*hyper((1/2 + 1/(2*(k*n + 1)),), (3/2, 3/2 + 1/(2*(k*n + 1))), -x**2*x**(2*k*n)/4)/(2*k*n*gamma(3/2 + 1/(2*(k*n + 1))) + 2*gamma(3/2 + 1/(2*(k*n + 1)))), True))
```

whereas with solve:

```
>>> integrate(sin(x**(n*k + 1)), x)
Piecewise((x*sin(1), Eq(k, -1/n)), (x*x**(k*n)*x**(k*n/(k*n + 1))*x**(1/(k*n + 1))*gamma(1/2 + 1/(2*(k*n + 1)))*hyper((1/2 + 1/(2*(k*n + 1)),), (3/2, 3/2 + 1/(2*(k*n + 1))), -x**2*x**(2*k*n)/4)/(2*k*n*gamma(3/2 + 1/(2*(k*n + 1))) + 2*gamma(3/2 + 1/(2*(k*n + 1)))), True))
```

I think heurisch is doing something similar with solve here:
https://github.com/sympy/sympy/blob/a44cc429c0c38e9abf6e60d421a94bb2f1de26c9/sympy/integrals/heurisch.py#L146

asmeurer replied:
I see. It needs to call solve so that it can get a chance to actually evaluate the integral.

asmeurer commented on sympy/integrals/meijerint.py:
```diff
@@ -1660,6 +1661,18 @@ def _meijerint_indefinite_1(f, x):
+            if not den.is_Number and not den.is_nonzero:
+                zero_sol = solve(den, dict=True)
+                for sol in zero_sol:
+                    zero_res = _meijerint_indefinite_1(f.subs(sol), x)
```
Does this internal function return an Integral object if it can't evaluate it?

ehren replied:
`_meijerint_indefinite_1` and `meijerint_indefinite` both return `None` on failure. But just calling `integrate` bypasses the problem with `meijerint_indefinite(x**n, x)`!

ehren replied:
turned out leaving it unevaluated until later was the better option than recursive integrate (see third commit message).

ehren replied:
I've got code that won't break the testsuite, only adding more special cases, but it turns out that just handling the division in the substitution in meijerint_indefinite_1 isn't adequate for treating all potential zero divisions. The better approach is sharing the heurisch logic. Will post that as a separate pull plus more info why the approach here can't be complete.

asmeurer replied:
That's great. I think there is also piecewise handling code in risch and manualintegrate, and possibly others, so it would be good to investigate those as well.

asmeurer commented on sympy/integrals/meijerint.py:
```diff
@@ -1660,6 +1661,18 @@ def _meijerint_indefinite_1(f, x):
+        # handle zero denominators in the above substitution
+        special_results = []
+        for den in (a, b):
+            if not den.is_Number and not den.is_nonzero:
```
You should use `is_zero`. "nonzero" means "real and not zero"

asmeurer commented:
I haven't tested this yet, but it looks like the right idea. The calling function calls piecewise_fold, so it shouldn't matter if the piecewise returned here is denested.

ehren committed:
review comments:

-use is_zero is not False instead of not is_nonzero
-use recursive call to integrate which allows an unevaluated result and bypasses problems when meijerint_indefinite can't handle the special case

ehren committed:
Instead of calling integrate recursively in _meijerint_indefinite_1, use
an unevaluated Integral for the special condition leaving it to the higher
level to evaluate. This bypasses the need to unpolarify the special
condition integrand (as the final meijerg result will be unpolarified by
existing code) and better handles integrate's kwargs; see newly added
testcase. Commit also unpolarifies the special case Piecewise's condition
similarly to other Piecewise handling in meijerint.

asmeurer commented:
> I suppose the problem is log(x) is not hypergeometric even though log(1+x) is.

Is this still an issue using Integral recursively instead of _meijerint_indefinite_1? The meijerg integrator has code to try to detect shifts that make things integrable, although I don't know if it can handle this case.

ehren committed:
changes necessary to keep testsuite working:

-return special case separately, merge with _meijerint_indefinite_1
result as last step (see pull comments)
-switch back to eagerly computing special case via integrate but ensure
integrand is unpolarified prior to attempting computation

also fixes a typo in a comment from previous commit

ehren commented:
> Is this still an issue using Integral recursively instead of _meijerint_indefinite_1? The meijerg integrator has code to try to detect shifts that make things integrable, although I don't know if it can handle this case.

There's no longer a problem now that integrate is used recursively.

Also without answering whether meijerint could/should handle the `integrate(x**n, x)` case it does handle (other than at x=1):

```
>>> integrate(1/x, x, meijerg=True)
Piecewise((log(x), Abs(x) < 1), (-log(1/x), 1/Abs(x) < 1), (-meijerg(((), (1, 1)), ((0, 0), ()), x) + meijerg(((1, 1), ()), ((), (0, 0)), x), True))
```

but the log(x) pops out from the residue handling (of the poles of the meijerg representation of the integrand?) here:
https://github.com/sympy/sympy/blob/87a885b56e57d99ef929dfbd32226923fe4f163f/sympy/simplify/hyperexpand.py#L2333

and not the hypergeometric term handling below that.

ehren commented:
Ultimately this pull has to be abandoned but here's an explanation for the last 4th commit which will pass the testsuite.

In the 3rd commit we had:

```
>>> x, a, b = symbols("x, a, b")
>>> integrate(1/x/(a + b*x)**Rational(1, 3), x, meijerg=True)
Piecewise((log(x)/a**(1/3), Eq(b, 0)), (-gamma(1/3)*hyper((1/3, 1/3), (4/3,), a*exp_polar(I*pi)/(b*x))/(b**(1/3)*x**(1/3)*gamma(4/3)), True))
```

but with this latest commit where the special case is returned as a separate tuple element (instead of merged immediately into the result computed by _meijerint_indefinite_1) we have

```
>>> integrate(1/x/(a + b*x)**Rational(1, 3), x, meijerg=True)
Piecewise((log(x)/a**(1/3), Eq(b, 0)), (2*log(1 - b**(1/3)*(a/b + x)**(1/3)/a**(1/3))*gamma(2/3)/(3*a**(1/3)*gamma(5/3)) + 2*exp(2*I*pi/3)*log(1 - b**(1/3)*(a/b + x)**(1/3)*exp_polar(2*I*pi/3)/a**(1/3))*gamma(2/3)/(3*a**(1/3)*gamma(5/3)) + 2*exp(-2*I*pi/3)*log(1 - b**(1/3)*(a/b + x)**(1/3)*exp_polar(4*I*pi/3)/a**(1/3))*gamma(2/3)/(3*a**(1/3)*gamma(5/3)), True))
```

which passes the test here https://github.com/sympy/sympy/blob/87a885b56e57d99ef929dfbd32226923fe4f163f/sympy/integrals/tests/test_meijerint.py#L662

Returning the special case separately is needed because of the check:

https://github.com/sympy/sympy/blob/87a885b56e57d99ef929dfbd32226923fe4f163f/sympy/integrals/meijerint.py#L1625

(the call `_has(res, hyper)` will return False if only one of the pieces has `hyper` leading to the testcase failure). This also indicates the meijerg code that already produces convergence conditions (e.g. _rewrite1) can't be easily modified to handle the special cases from poles in the denominator (at least not without modifying the `_has` acceptance logic).

In addition, if we used the logic "use the special case (arising from the denominator in the substitution in _meijerint_indefinite_1) that corresponds to the same integrand used to compute the ultimately accepted result of _meijerint_indefinite_1 (that is which uses the same splitting point substitution) we'd get a nonsensical special case:

```
>>> integrate(1/x/(a + b*x)**Rational(1, 3), x, meijerg=True)
Piecewise((log(zoo*a**(4/3) + a**(1/3)*x)/a**(1/3), Eq(b, 0)), (2*log(1 - b**(1/3)*(a/b + x)**(1/3)/a**(1/3))*gamma(2/3)/(3*a**(1/3)*gamma(5/3)) + 2*exp(2*I*pi/3)*log(1 - b**(1/3)*(a/b + x)**(1/3)*exp_polar(2*I*pi/3)/a**(1/3))*gamma(2/3)/(3*a**(1/3)*gamma(5/3)) + 2*exp(-2*I*pi/3)*log(1 - b**(1/3)*(a/b + x)**(1/3)*exp_polar(4*I*pi/3)/a**(1/3))*gamma(2/3)/(3*a**(1/3)*gamma(5/3)), True))
```

(see the comment in the 4th commit)

Anyhow, looking at the desired result with gamma it's apparent that this latest 4th commit also misses the pole `a = 0` (because it doesn't arise from the substitution in _meijerint_indefinite_1). For this the better change is to reuse the heurisch_wrapper logic.

ehren commented:
Tests passed other than the known issue on master but closing in favour of #20270.
